rfcs/0133-ipfs.md

```diff
@@ -0,0 +1,284 @@
+---
+feature: ipfs
+start-date: (fill me in with today's date, YYYY-MM-DD)
+author: John Ericsion (@Ericson2314) on behalf of [Obsidian Systems](https://obsidian.systems)
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Integrate Nix with IPFS, in phases of increasing sophistication.
+This follows the work done and described in https://github.com/obsidiansystems/ipfs-nix-guide/ .
+
+# Motivation
+[motivation]: #motivation
+
+## Binary distribution
+
+Currently distributing Nix binaries takes a lot of bandwidth and storage.
+This is a barrier to being a Nix user in areas of slower internet --- which includes the vast majority of the world's population at this time.
+This is also a barrier to users running their own caches.
+
+Content-addressing opens up a *huge* design space of solutions to get around such problems.
+IPFS explores many of those solutions.
+
+## Source distribution and archival
+
+A goal of the Nix ecosystem is to package software in a way that never bitrots.
+Getting in the way of that, however, is the fact source code frequently goes off-line.
+The Software Heritage archive is the best in the world, and a natural partner in this effort.
+
+Unfortunately, as https://www.tweag.io/blog/2020-06-18-software-heritage/ describes at the end, a major challenge is the way nix content-addresses software.
+First of all, Nix hashes sources in bespoke ways that no other project will adopt.
+Second of all, tarballs instead of the underlying files leaking non-normative details (compression, odd perms, etc.).
+
+We should natively support git file hashing, which Git repos and Software Heritage both support.
+This will completely obliterate these issues.
```

Comment:
> Or does the lack of adoption of SHA2 in Git pose issues.

I think it does. I am happy to prioritize adding SHA-2 Git (and SWHID) content-addressing ahead of wide adoption, and indeed try to lead that change, but I am hesitant to get rid of SHA-1 support until that change happens.

Put another way, as long as Git SHA-1 is omnipresent, our our additional hashes adding much value? Whoever wrote down the additional hashes was verifying something by SHA-1, so a hash collision attack could sneak in then.

Reply:
> Or does the lack of adoption of SHA2 in Git pose issues.

I think it does. I am happy to prioritize adding SHA-2 Git (and SWHID) content-addressing ahead of wide adoption, and indeed try to lead that change, but I am hesitant to get rid of SHA-1 support until that change happens.

Put another way, as long as Git SHA-1 is omnipresent, our our additional hashes adding much value? Whoever wrote down the additional hashes was verifying something by SHA-1, so a hash collision attack could sneak in then.

Reply:
I think I am partly confused because it seems like there may be multiple uses of the hash, both for the native importing of sources that use Git (which the sha1 support is definitely a huge benefit) and use for non-git sources and other artifacts (where the benefits are less clear). I think if we can make it clear which hash is used for which things and where parity matters then I think you are right and we will need to support sha1-git for the time being.

Reply:
We could I suppose *only* allow git sha-1 fetching with `builtins.fetchgit` and friends, but what about fixed-output derivations like Nixpkgs uses that also fetch git repos?

One thing I am hoping we can get to is just writing down *one* hash for git repo fetching, and if it is a commit hash or tag object instead of git hash, using a [Merkle inclusion proof](https://computersciencewiki.org/index.php/Merkle_proof) to connect it to the tree hash we'd ultimately use.

That would be a *huge* ergonomics improvement for Nix.

Reply:
I'm trying to comprehend the security implications of sha1 hashes for store content and I think they aren't as bad as I thought. Since I think that sha1 hashes in the store will look different (different length) than other hashes you won't be able to replace a non-sha1 path with a colliding sha1 path. So I think it will only reduce security for places where it is used which is probably not that much of a problem. With little thought the worst attack I can think of is forcing someone to pull your source for a build. That is obviously bad but as sha1 is (hopefully) phased out less and less packages will be affected by this, just having sha1 as a supported hash doesn't reduce the integrity of derivations that don't use it.

That being said I wonder if it still makes sense to have `builtins.fetchgit` still support providing an additional hash. This would be less ergonomic than just providing one hash for both commit ID + fixed output verification but would provide higher confidence in the integrity of the downloaded result. So this would basically look like downloading from IPFS via the sha1 git hash and then verifying the result again with the other hash.

Reply:
> So I think it will only reduce security for places where it is used which is probably not that much of a problem.

Yup. And the name of the hash is always included in the store path hash computation too. And the *second* round of hashing used for that is SHA-256.

I should add some text here emphasizing this property.

Reply:
> That being said I wonder if it still makes sense to have `builtins.fetchgit` still support providing an additional hash. This would be less ergonomic than just providing one hash for both commit ID + fixed output verification but would provide higher confidence in the integrity of the downloaded result.

Yeah that is just what we have today, more or less.

> So this would basically look like downloading from IPFS via the sha1 git hash and then verifying the result again with the other hash.

This is a little tricky because right now one *just* fetches from IPFS by using it as a substitutor, this is no `builtins.fetchipfs` because no libfetchers additional logic is needed! But there are still options we can discuss.

Reply:
A simple option to work around this would be to keep `fetchgit` as a sha1 derivation but then wrap it in another derivation that just checks the hash and copies the result. You wouldn't be able to download this derivation from IPFS but it is likely cheap enough that it isn't a huge issue to rebuild it locally.

Reply:
> Git hashing is currently on SHA1 which is showing weaknesses.

There is some useful and recent information on the git-sha256 situation [in this LWN article](https://lwn.net/Articles/898522/).

Reply:
Thanks! My take away is:

1. Make sure we also support [SHA-1DC](https://github.com/cr-marcstevens/sha1collisiondetection) not vanilla SHA-1.
2. Try out supporting greenfield SHA-256 git repos.
3. Do not worry about migrating code as upstream git doesn't have a plan to migrate / dual-sign repos either.

We didn't do (2) in 2020 when the support in git itself was new, but we could do (2) now.

Ultimately crypto-agility is an interesting and important problem, and I think there is a great opportunity to lead the way here *eventually*, but lets worry about trying to catch up to the state of the art before we push it forward!

Reply:
So the git upload is signed by the signing key to make sure that it is trusted?

Reply:
> 3\. dual-sign

I think Ericson2314 meant "dual-hash".

Reply:
How would dual hashing prevent the following scenario:
- a malicious commit is produced with a sha1 collision
- it gets uploaded to ipfs before the nix CI can process it

Also how would it be prevented that wrongly signed nix store entries are uploaded to ipfs? Would the hash then be different or could a collision be produced?

Reply:
A proper dual hashing solution would be something like `sha256 || sha1` hashes so you have ways to verify not only the root object but every object in it's closure.

Reply:
Collapsing IPFS threads. Since the RFC is now about Git hashing, those drawbacks or more intrinsic to the goal. (Whereas with IPFS we could hash in other way.s)